agent/consul/leader_connect.go

```diff
@@ -173,6 +177,9 @@ func (s *Server) initializeCA() error {
 	if err != nil {
 		return err
 	}
+
+	s.caProviderReconfigurationLock.Lock()
+	defer s.caProviderReconfigurationLock.Unlock()
 	s.setCAProvider(provider, nil)
```

Comment:
Might be a good idea to document in the comment for `setCAProvider` that it is often called while holding `caProviderReconfigurationLock` and so must never take that or call anything that does.

I think it's unlikely it would but it's not super obvious that would cause a deadlock which might bite us eventually.

Reply:
I added that warning in a bunch of places.